src/attributes.rs

```diff
@@ -0,0 +1,235 @@
+use std::rc::Rc;
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Attribute<'a> {
+    pub is_inner: bool,
+    pub content: Rc<AttributeValue<'a>>,
+}
+
+impl<'a> Attribute<'a> {
+    pub fn as_string(&self) -> String {
+        format!(
+            "#{}[{}]",
+            if self.is_inner { "!" } else { "" },
+            self.content.raw_value
+        )
+    }
+
+    pub fn new(raw: &'a str) -> Self {
+        let raw_without_closing = raw
+            .trim()
+            .strip_suffix(']')
+            .expect("Attribute has to be closed by a square bracket.");
+
+        if let Some(raw_content) = raw_without_closing.strip_prefix("#[") {
+            Attribute {
+                is_inner: false,
+                content: Rc::new(AttributeValue::new(raw_content)),
+            }
+        } else if let Some(raw_content) = raw_without_closing.strip_prefix("#![") {
+            Attribute {
+                is_inner: true,
+                content: Rc::new(AttributeValue::new(raw_content)),
+            }
+        } else {
+            panic!("Attribute has to start either with `#[` or `#![`.");
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
```

Comment:
Any particular reason this isn't also `Eq`?

Reply:
No, I just didn't derive it, because I need only PartialEq. Is it better, in general, to derive both if possible?

Reply:
Yes, if the type is able to satisfy `Eq` then it's best to derive it. There's no downside, only upside.

You probably have heard of "[Chesterton's Fence](https://wiki.lesswrong.com/wiki/Chesterton%27s_Fence)": the principle that if something looks strange, it's likely that way for a reason even if the reason may be non-obvious, and that this reason must be understood before the system is modified. In this situation, we have a corollary of Chesterton's Fence: don't design things in an unusual way when there's no reason for that. If I was reading some code and saw a type that is `PartialEq` but isn't `Eq`, I'd assume that's for some non-obvious reason and would try to figure out why `Eq` might not be possible to satisfy.

This corollary of Chesterton's Fence is sometimes also called the Principle of Least Astonishment: https://en.wikipedia.org/wiki/Principle_of_least_astonishment